### PR TITLE
Feature/infinite scroll

### PR DESCRIPTION
**Pull request by NathanVss**
- [x] Algorithm implementations
- [x] React-query cache
- [x] Infinite scroll displaying real enrollments and orders
- [x] Infinite scroll without react dependency
- [x] Empty placeholder
- [x] Errors banner
- [x] Disable future requests if reached EOF
- [x] Test edge cases in real case ( items = 0, items < chunkSize ) 
- [x] Testing (Jest)
- [x] Make sure `setQueryData` sets the same `staleTime` as `useQuery`

### CODE REVIEW

**Review by rlecellier — COMMENTED** [review]
That's a nice array party.

File: src/frontend/js/pages/DashboardCourses/index.tsx [hunk 1]
```diff
@@ -1,87 +1,83 @@
...
-              ),
+        <>
+          {count === 0 && (
+            <p className="dashboard__courses__empty">
+              <FormattedMessage {...messages.emptyList} />
+            </p>
+          )}
+          <div className="dashboard__courses__list">
+            {data.map((datum) => (
+              <div key={datum.item.id} className="dashboard__courses__list__item">
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/index.tsx:53]: uuid are supposed to be universe-unique 🤔

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+const messages = defineMessages({
+  errorGet: {
+    id: 'hooks.useOrdersEnrollments.errorGet',
+    description: 'Error message shown to the user when orders or enrollments fetch request fails.',
+    defaultMessage: 'An error occurred while fetching orders and enrollments. Please retry later.',
+  },
+});
+
+const checkIntegrity = (previousStack: Data[], newStack: Data[], integrityIndex: number) => {
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:29]: I think `index` is more appropriate because it refers to an array position, where `count` is more abstract imo

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  if (previousStack.length < integrityIndex || newStack.length < integrityIndex) {
+    return false;
+  }
+
+  // If we find a difference, we return false.
+  return !newStack.slice(0, integrityIndex).find((item, index) => {
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:35]: Keeping `index` would avoid doing this kind of modification, and also to keep track of small mathematics here and there IMO 😕

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    return previousStack[index].item.id !== item.item.id;
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:36]: 1. That's maybe caused to the ` - 1` included in the previous suggestion. Anyway, let's talk about this in more details during the daily meeting. 

2. Totally! I didn't provided it because it's the default order, but you're right, I"ll add it :)

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    _stack.sort((a, b) => {
+      const aDate = new Date(a.item.created_on);
+      const bDate = new Date(b.item.created_on);
+      return bDate.getTime() - aDate.getTime();
+    });
+
+    return { entities, totalCount };
+  };
+
+  return { fetchEntitiesInStack };
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:147]: That would break the algorithm. Here we want to add in place the non existing entities. Where your code returns the non existing ones. Furthermore you remove the sorting which is the key point to guarantee integrity. but maybe I get the point wrong, let's talk about it during daily.

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    let currentStack: Data[];
+    let newTotalCount;
+    while (true) {
+      pageToLoad += 1;
+      currentStack = [...previousStack];
+      log('Fetching page ' + pageToLoad + ' ...');
+
+      try {
+        // eslint-disable-next-line no-await-in-loop
+        const { totalCount: tmpTotalCount } = await fetchEntitiesInStack(pageToLoad, currentStack);
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:204]: ditto related to my last comment

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+        if (typeof newTotalCount === 'undefined') {
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:205]: Hmmm, good catch. I need to work it out.

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+        // Having the same length means that we reached the end of the list.
+        if (currentStack.length === previousStack.length) {
+          setIntegrityCount(currentStack.length);
+          break;
+        }
+        if (checkIntegrity(previousStack, currentStack, wantedIntegrityCount)) {
+          setIntegrityCount(wantedIntegrityCount);
+          break;
+        }
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:226]: Nope! `currentStack` also includes all the entities fetched from the start. ( To guarantee integrity we merge then re-sort everything in case where a newly fetched entity would be thrown in front of the array )

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+        break;
+      }
+    }
+
+    log('FINAL STACK', currentStack, 'totalCount', totalCount, 'newTotalCount', newTotalCount);
+    if (typeof totalCount === 'undefined') {
+      setTotalCount(newTotalCount);
+    }
+    setPage(pageToLoad);
+    setStack(currentStack);
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:239]: We must sort before the integrity checking. 😕

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    setIsSyncing(false);
+  };
+
+  useEffect(() => {
+    if (cursor > integrityCount) {
+      syncIntegrityCount(cursor);
+    }
+  }, [cursor]);
+
+  const cursorToUse = Math.min(cursor, integrityCount);
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:249]: During the syncing ( aka fetching the next slice of entities ), we would always have `cursor > integrityCount`, so we can only guarantee the integrity of the minimum of both.

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+          'newTotalCount',
+          newTotalCount,
+        );
+
+        // Having the same length means that we reached the end of the list.
+        if (currentStack.length === previousStack.length) {
+          setIntegrityCount(currentStack.length);
+          break;
+        }
+        if (checkIntegrity(previousStack, currentStack, wantedIntegrityCount)) {
```
Comment (jbpenrath) [inline src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:223]:
```suggestion
        if (hasIntegrity(previousStack, currentStack, wantedIntegrityCount)) {
```

Comment (jbpenrath) [inline src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:223]: You should wrap in the `try...catch` only the statement that could raise the error.

File: src/frontend/js/hooks/useOrders.ts [hunk 12]
```diff
@@ -10,10 +10,12 @@ import {
   UseResourcesProps,
 } from './useResources';
 
-type OrderResourcesQuery = ResourcesQuery & {
+export type OrderResourcesQuery = ResourcesQuery & {
   course?: Course['code'];
   product?: Product['id'];
   state?: OrderState[];
+  page?: number;
+  page_size?: number;
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/hooks/useOrders.ts:18]: Nope, I'm adding it.

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+    if (!res.next) {
+      eof.current = { ...eof.current, [queryKeyString]: filters.page };
+      // Eof is cached based, the same way, we cache the fetching data. Otherwise it there would
+      // be request to non existing pages after reload.
+      queryClient.setQueryData(EOF_QUERY_KEY, eof.current);
+    }
+    return res;
+  };
+
+  const fetchEntities = async (pageToLoad: number) => {
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:93]: Why not :)

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+      }
+    });
+    // Sort according to created_on
+    _stack.sort((a, b) => {
+      const aDate = new Date(a.item.created_on);
+      const bDate = new Date(b.item.created_on);
+      return bDate.getTime() - aDate.getTime();
+    });
+
+    return { entities, totalCount };
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:144]: To be correct it corresponds to the count of all entities in database. Does this information changes your comment?

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+      entities = [
+        ...entities,
+        ...enrollmentsResponse.results.map((enrollment) => ({
+          type: DataType.ENROLLMENT,
+          item: enrollment,
+        })),
+      ];
+      totalCount += enrollmentsResponse.count;
+    }
+
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx:124]: Good idea! I used `Promise.all` instead because we want it to throw in case of failure to reach the try catch of the main while(true) loop.

File: src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx]: Oh ok! Yup looks fine now.

**Review by jbpenrath — APPROVED** [review]
LGTM ✅
